yazi-plugin/src/external/highlighter.rs

```diff
@@ -56,51 +59,238 @@ impl Highlighter {
 		syntaxes.find_syntax_by_first_line(&line).ok_or_else(|| anyhow!("No syntax found"))
 	}
 
-	pub async fn highlight(&self, skip: usize, limit: usize) -> Result<Text<'static>, PeekError> {
-		let mut reader = BufReader::new(File::open(&self.path).await?);
+	pub async fn highlight(
+		&self,
+		skip: usize,
+		area: Rect,
+		wrap: bool,
+	) -> Result<Text<'static>, PeekError> {
+		let reader = BufReader::new(File::open(&self.path).await?);
 
 		let syntax = Self::find_syntax(&self.path).await;
 		let mut plain = syntax.is_err();
 
-		let mut before = Vec::with_capacity(if plain { 0 } else { skip });
-		let mut after = Vec::with_capacity(limit);
+		let (before, after): (Vec<String>, Vec<String>);
+		if wrap {
+			(before, after) = Self::before_after_wrapped(skip, area, &mut plain, reader).await?;
+		} else {
+			(before, after) = Self::before_after(skip, area, &mut plain, reader).await?;
+		}
 
-		let mut i = 0;
+		if plain {
+			if !wrap {
+				let indent = " ".repeat(PREVIEW.tab_size as usize);
+				return Ok(Text::from(after.join("").replace('\t', &indent)));
+			}
+			Ok(Text::from(after.join("")))
+		} else {
+			Self::highlight_with(before, after, syntax.unwrap()).await
+		}
+	}
+
+	async fn before_after_wrapped(
+		skip: usize,
+		area: Rect,
+		plain: &mut bool,
+		mut reader: BufReader<File>,
+	) -> Result<(Vec<String>, Vec<String>), PeekError> {
+		let mut before = Vec::with_capacity(if *plain { 0 } else { skip });
+		let mut after = Vec::with_capacity(area.height as usize);
+
+		let mut long_lines = vec![];
 		let mut buf = vec![];
+		// If we want to indent plain text, we have to decide if it is plain
 		while reader.read_until(b'\n', &mut buf).await.is_ok() {
```

Comment:
Why do we need both `while` and `for` loops here? It seems like `while` will keep reading until the end of the file without stopping early, I think we should exit the loop as soon as we have enough lines for preview. 

Is it possible to move `if long_line.is_empty() || lines_handled > skip + area.height as usize {` up into the `while` loop and combine the `while` and `for` into one loop?

Reply:
At first I thought we have to go line by line twice, but now it seems it's possible to do it once. I tried to refactor code according to this. Now we traverse a file once, like before